src/main/scala/sigmastate/serialization/ValueSerializer.scala

```diff
@@ -107,16 +109,28 @@ object ValueSerializer extends SigmaSerializerCompanion[Value[SType]] {
       getSerializer(opCode).asInstanceOf[ValueSerializer[v.type]].serializeBody(v, w)
   }
 
+  private val nestedValuesDepthPerReader = TrieMap[Int, Int]()
```

Comment:
@ergomorphic I found it super ugly, but the alternative (passing the depth through `parseBody`) seems even uglier. I'm open to any ideas. :)

Reply:
@greenhat We already passing Reader along the tree (i.e. along all recursive calls). This means we can add  `getLevel` `setLevel` methods in the Reader interface so that serializers can update the corresponding state field if they need it.

Reply:
@ergomorphic I've ruled that out initially due to probable SRP violation, but it's least of all other evils. :) Thanks.

Reply:
@ergomorphic Done `get/setLevel`.